Introduce JavaScriptEvaluationResult

achristensen07 opened the pull request:
#### 855e48e2781f6a309352a8afe46fa02786c14a42
<pre>
Introduce JavaScriptEvaluationResult
<a href="https://bugs.webkit.org/show_bug.cgi?id=288698">https://bugs.webkit.org/show_bug.cgi?id=288698</a>

Reviewed by Sihui Liu.

Cocoa platforms only allow 6 types to come out of WKWebView.evaluateJavaScript:
NSNumber, NSString, NSDate, NSArray, NSDictionary, and NSNull.
Currently, we serialize SerializedScriptValue::wireBytes then deserialize them in the UI
process with the aid of a JSContext then convert them to these ObjC types.
This introduces an abstraction that represents what is currently a std::span&lt;const uint8_t&gt;
on all platforms, but it will enable us to make it a std::variant of 6 types on platforms
that only need those.  That will allow us to simplify the IPC format and deserialize without
the need of a JSContext.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp: Copied from Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h.
(WebKit::JavaScriptEvaluationResult::serializedScriptValue const):
(WebKit::JavaScriptEvaluationResult::toWK const):
(IPC::std::optional&lt;WebCore::ExceptionDetails&gt;&gt;&gt;::create):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h: Copied from Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h.
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::wireBytes const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm: Copied from Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h.
(WebKit::JavaScriptEvaluationResult::toID const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageEvaluateJavaScriptInMainFrame):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _evaluateJavaScript:asAsyncFunction:withSourceURL:withArguments:forceUserGesture:inFrame:inWorld:completionHandler:]):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRunJavaScriptWithParams):
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::runScript):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::runJavaScriptInMainFrame):
(WebKit::WebPageProxy::runJavaScriptInFrameInScriptWorld):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):
(WebKit::WebPage::runJavaScriptInFrameInScriptWorld):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/291252@main">https://commits.webkit.org/291252@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d8a4531e65158d2c0c4e216678ff53f493e4a44

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92430 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11971 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97414 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42937 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12249 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20432 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/28288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95432 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9309 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83709 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51154 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9010 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42268 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99440 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19481 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19731 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79587 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79108 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23652 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14714 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19465 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19152 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22612 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20892 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai reviewed (approved):
So this just introduces the wrapper class but not actually doing the filtering; seems a good first step!
Please fix the tests before landing -- looks like it's caused by returning NSError of unexpected JavaScriptEvaluationResult; for compatibility maybe we should keep the old behavior.

achristensen07 committed:
Introduce JavaScriptEvaluationResult
https://bugs.webkit.org/show_bug.cgi?id=288698

Reviewed by Sihui Liu.

Cocoa platforms only allow 6 types to come out of WKWebView.evaluateJavaScript:
NSNumber, NSString, NSDate, NSArray, NSDictionary, and NSNull.
Currently, we serialize SerializedScriptValue::wireBytes then deserialize them in the UI
process with the aid of a JSContext then convert them to these ObjC types.
This introduces an abstraction that represents what is currently a std::span<const uint8_t>
on all platforms, but it will enable us to make it a std::variant of 6 types on platforms
that only need those.  That will allow us to simplify the IPC format and deserialize without
the need of a JSContext.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp: Copied from Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h.
(WebKit::JavaScriptEvaluationResult::serializedScriptValue const):
(WebKit::JavaScriptEvaluationResult::toWK const):
(IPC::std::optional<WebCore::ExceptionDetails>>>::create):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h: Copied from Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h.
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::wireBytes const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm: Copied from Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h.
(WebKit::JavaScriptEvaluationResult::toID const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageEvaluateJavaScriptInMainFrame):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _evaluateJavaScript:asAsyncFunction:withSourceURL:withArguments:forceUserGesture:inFrame:inWorld:completionHandler:]):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRunJavaScriptWithParams):
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
(WebKit::RemoteInspectorProtocolHandler::runScript):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::runJavaScriptInMainFrame):
(WebKit::WebPageProxy::runJavaScriptInFrameInScriptWorld):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):
(WebKit::WebPage::runJavaScriptInFrameInScriptWorld):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/291252@main
